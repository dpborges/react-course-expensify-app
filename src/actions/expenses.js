import uuid from 'uuid';
import database from '../firebase/firebase';


// ADD_EXPENSE
export const addExpense = (expense) => ({
    type: 'ADD_EXPENSE',
    expense
});

// Function used by middleware redux-thunk
// this function takes expense data as an argument and returns a function that takes 
// dispatch as an argument while expense data is already been bound to the function.
export const startAddExpense = (expenseData = {}) => {
    return (dispatch) => {
        // the const below is using destructuring to assign properties of expenseData 
        // object to the individual fields. The fields are then used below to construct
        // the expense object.
        const {
            description = '', 
            note = '', 
            amount = 0, 
            createdAt = 0 
        } = expenseData;
        const expense = { description, note, amount, createdAt };

        // add expense to firebase
        return database.ref('expenses').push(expense).then((ref) => {
            // update redux store
            dispatch(addExpense({
                id: ref.key,
                ...expense
            }));
        });
    ;}
};

// REMOVE_EXPENSE
export const removeExpense = ( { id } = {}) => ({
    type: 'REMOVE_EXPENSE',
    id
});

// EDIT_EXPENSE
export const editExpense = (id, updates) => ({
    type: 'EDIT_EXPENSE',
    id,
    updates
});

//SET_EXPENSES; passes in array of all expenses in our fixture. This is called before
//              each test case to initiale expenses.
export const setExpenses = (expenses) => ({
    type: 'SET_EXPENSES',
    expenses
});


// Function used by redux-thunk middleware.
// this function fetches data from firebase, parses data into an array, and dispatches
// SET_EXPENSES
export const startSetExpenses = () => {
    return (dispatch) => {
        // fetch data from firebase
        const expenses = [];
        return database.ref('expenses').once('value').then((snapshot) => {
            // parse data into an array
            snapshot.forEach((childSnapshot) => {
                expenses.push({
                    id: childSnapshot.key,
                    ...childSnapshot.val()
                });
            });
            console.log('array from database ', expenses);
            // update redux store with data fetched from database
            dispatch(setExpenses(expenses));
        });
    ;}
};
