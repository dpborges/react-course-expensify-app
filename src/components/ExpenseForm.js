import React from 'react';
import moment from 'moment';
import { SingleDatePicker } from 'react-dates';
// Fort testing purpose, the import below was moved to app.js
// MOVED to App.js ==> import 'react-dates/lib/css/_datepicker.css';


// Goal is to keep track of local state for all of the inputs. Only when form is sumbitted
// do we want to submit change to redux store. That submission is handled by parent component.
// The form here is simply responsible for presenting the form, validating inputs, saving data in 
// local state, and passing data to function prop for dispatch. As you may notice, there is
// is no connect() function imported for this component, therefore it cannot connect to redux.

export default class ExpenseForm extends React. Component {
    constructor(props) {
        super(props);
    
        // local component state
        this.state = {
            description: props.expense ? props.expense.description : '',
            note:     props.expense ? props.expense.note : '',
            amount:   props.expense ? (props.expense.amount / 100).toString() : '',
            createdAt:  props.expense ? moment(props.expense.createdAt) : moment(),
            calendarFocused: false,
            error: ''
        };
    }

    // Be aware that "e.target.value" cannot be used in the callback inside this.setState()
    // You may used e.persist() instead of the assignment line before setState()
    onDescriptionChange = (e) => {
        const description = e.target.value;     // assign input value to description; allows to use object shorthand on next line
        this.setState(() => ({ description })); // update description on local state using ES6 shorthand
    };

    onNoteChange = (e) => {
        const note = e.target.value;     // assign input value to description; allows to use object shorthand on next line
        this.setState(() => ({ note })); // update description on local state using ES6 shorthand
    };

    onAmountChange = (e) => {
        const amount = e.target.value;
        if (!amount || amount.match(/^\d{1,}(\.\d{0,2})?$/)) {
            this.setState(() => ({ amount }));
        }
    };

    onDateChange = (createdAt) => { 
        if  (createdAt) {
            this.setState(() => ({ createdAt }));
        }
    };
    
    onFocusChange = ({ focused }) => {
        this.setState(() => ({ calendarFocused: focused }));    
    };

    // Note: this local onSubmit method does validation but does not dispatch the 
    //       action (eg add or edit). The dispatch is done by the onSubmit method
    //       passed in as a function prop, from the AddExpensePage. That function prop
    //       is called (this.props.onSubmit) with data needed, then action is dispatched
    //       in the parent component (AddExpensePage).
    onSubmit = (e) => {
        e.preventDefault();
        if (!this.state.description || !this.state.amount) {
            this.setState(() => ({ error: 'Please provide description and amount'}))
        } else {
            this.setState(() => ({ error: ''}))
            // Call "function prop" passed in from AddExpensePage.js
            this.props.onSubmit({
                description: this.state.description,
                amount: parseFloat(this.state.amount, 10) * 100, // convert text to number
                createdAt: this.state.createdAt.valueOf(),
                note: this.state.note
            })
        }
    };
    
    render() {
        return (
            <form className="form" onSubmit={this.onSubmit}>
            {this.state.error && <p className="form__error">{this.state.error}</p>}
                <input 
                    type="text"
                    placeholder="Description"
                    autoFocus
                    className="text-input"
                    onChange={this.onDescriptionChange}  // change the state.description value
                    value={this.state.description}    // render component with value from state object
                />
                <input 
                    type="text"
                    placeholder="Amount"
                    className="text-input"
                    value={this.state.amount}
                    onChange={this.onAmountChange}
                />
                <SingleDatePicker
                    date={this.state.createdAt}
                    onDateChange={this.onDateChange}
                    focused={this.state.calendarFocused}
                    onFocusChange={this.onFocusChange}
                    numberOfMonths={1}
                    isOutsideRange={() => false }
                />
                <textarea
                    placeholder="Add a note for your expense (optional)"
                    className="textarea"
                    onChange={this.onNoteChange}    // change the state.description value
                    value={this.state.note}         // render component with value from state object
                >
                </textarea>
                <div>
                  <button className="button">Save Expense</button> 
                </div>
            </form>
        )
    }
}