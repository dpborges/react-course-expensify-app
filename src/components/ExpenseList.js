import React from 'react';
import { connect } from 'react-redux';
import ExpenseListitem from './ExpenseListItem';
import selectExpenses from '../selectors/expenses';

// SEE LESSONS 100 - 101

//  F I R S T   R E A C T - R E D U X  C O N N E C T E D  C O M P O N E N T

// Steps to connecting component to redux
// Step 1:  set up redux store Provider in the root of your applcation (eg. app.js)
// Step 2:  Create HOC using connect function provided by react-redux
// Step 3:  Define things you want to get off the store using mapStateToProps
// Step 4:  Specify presentation component you want to connect the redux store with.


//This export used by test file and default export used by application.
// by doing this, you disconnnect from the store and you can pass in props in test cases
export const ExpenseList = (props) => (        
    <div>                                      
        <h1>Expense list</h1>
        {
            props.expenses.length === 0 ? (
                <p>No expenses</p>
            ) : (
                props.expenses.map((expense) => {
                    return   <ExpenseListitem  key={expense.id}  {...expense} />;
                })
            )
        }
    </div>
);

// ====================================================================================
//                            Final React-Redux Impl
// ====================================================================================
// Note, once you connected a component to a redux store, every time state changes,
// your component will be re-rendered.
// To understand how compoments are connected to redux, I strongly suggest uncommenting
// lines between the CODE BLOCK line in section 
// "First React-Redux Impl" then come back and look at this section showing 
// "Final React-Redux Imp", as it may make more sense. What was done was to separate
// the connect function and assign it to mapStateToProps, then passing that function
// in to connect instead of having it in-line as shown in "First React-Redux Impl".


// Version # 1 of mapStateToProps - return expenses with no filtering
// const mapStateToProps = (state) => {    // Step 3
//     // N/V pairs returned below will be used to pass in as props to presentation component
//     return {
//         expenses: state.expenses,
//         filters: state.filters
//     };
// }

// Version # 2 of mapStateToProps - 
// Renders filtered list by applying selectExpenses in  our selectors folder.
// This function effectively overrides current state. Instead of returning property
// from the application state tree, it returns a filtered list of one of the properties.
const mapStateToProps = (state) => {    // Step 3
    return {
        expenses: selectExpenses(state.expenses, state.filters)
    };
};



// Here state is retured as props then passed into ExpenseList component.
export default connect(mapStateToProps)(ExpenseList);    // Step 2 & 4

// ====================================================================================
//                            First React-Redux Impl
// ====================================================================================
// The ConnectedExpenseList below is used as Wrapper function for our component, 
// hence why we export ConnectedExpenseList instead of ExpenseList.
// The connect()function,  is a curry function which returns a function, and hence why 
// its called  a second time with presentation component.
// The Object returned from connect's first function parameter is an object of name/value
// pairs that gets passed as props to the connected presentation component.
// You can return an object with arbitrary name/value pairs (like your name: 'dan'), or you
// can return properties from your redux store, like expenses and filters.
// **********************  CODE BLOCK **********************
// const ConnectedExpenseList = connect((state) => {

//     // N/V pairs returned below get passed in as props to presentation component
//     return {
//         expenses: state.expenses
//     };

// })(ExpenseList);

// export default ConnectedExpenseList;
// **********************  CODE BLOCK **********************