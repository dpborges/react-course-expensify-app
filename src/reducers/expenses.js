// Expenses Reducer
const expensesReducerDefaultState = []; // set default state
const expensesReducer = (state = expensesReducerDefaultState, action) => {
    switch (action.type) {
        case 'ADD_EXPENSE':
            return [...state, action.expense];    // returns state changes to redux
        case 'REMOVE_EXPENSE':
            console.log(`remove expense action called with id ${action.id}`);
            return state.filter(({ id }) =>  id !==  action.id )
        case 'EDIT_EXPENSE':
            return state.map((expense) => {
                if (expense.id === action.id) {
                    return {
                        ...expense,             // returns state changes to redux
                        ...action.updates       
                    }
                } else {
                    return expense;             // returns state changes to redux
                }
            })
        default: 
            return state;
    }
}

export default expensesReducer;