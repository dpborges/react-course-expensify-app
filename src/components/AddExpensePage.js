import React from 'react';
import { connect } from 'react-redux';
import ExpenseForm from './ExpenseForm';
import { addExpense } from '../actions/expenses'

// NOTE:  Be sure to read comments on the original AddExpensePage within AddExpenseOrig.js file.
// During lecture,  the component was converted from a functional component
// to a class component and an onSubmit function was added to the class. Below is the refactored component.
// Notice there is no call to dispatch in our component. The dispatch is within the MapDispatchToProps function
// which maps our dispatch to a function prop. 
// The last change made (for testing purposes) was to export the class, in order to test the unconnected version of
// the component.

export class AddExpensePage extends React.Component {

    onSubmit = (expense) => {
        this.props.addExpense(expense);
        this.props.history.push('/');
    };

    render() {
        return (
            <div>
                <h1>Add Expense </h1>
                <ExpenseForm 
                    // Passes the function prop "onSubmit" to the ExpenseForm component. 
                    // Note, function prop expects an expense object to be passed. 
                    onSubmit={this.onSubmit}                         
                />
            </div>
        )
    }
}


// The goal of mapDispatchToProps is to return an object that Maps your dispatch call to a prop vs calling dispatch inline.
// This allows you to abstract away dispatch from the compoment itself.
// Instead of having following inline code:    props.dispatch(addExpense(expense))    
// in the onSubmit prop used to render ExpenseForm above, we create the mapDispatchToProps function and 
// use following:                              props.onSubmit(expense)
// See the BEFORE and AFTER changes noted in the props of the  <ExpenseForm /> compoment above.
const mapDispatchToProps = (dispatch) => ({
        addExpense: (expense)  => dispatch(addExpense(expense))
});


export default connect(undefined, mapDispatchToProps)(AddExpensePage);