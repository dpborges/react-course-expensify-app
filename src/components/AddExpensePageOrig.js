import React from 'react';
import { connect } from 'react-redux';
import ExpenseForm from './ExpenseForm';
import { addExpense } from '../actions/expenses'



// Note original state of AddExpensePage before setting up our test used the code flagged as "line BEFORE setting up test case".
// The problem was that although we can easily have dispatch replaced by spy, the function addExpense(expense), 
// within the dispatch, has been imported.
// The way around this, is to make use of connect's "mapDispatchToProps", which is the 2nd
// parameter of react's connect functon. See comments below on mapDispatchToProps function. 
const AddExpensePage = (props) => (     // Props is here to be able to pass dispatch
    <div>
        <h1>Add Expense </h1>
        <ExpenseForm 
            // Passes the function prop "onSubmit" to the ExpenseForm component. 
            // Note, function prop expects an expense object to be passed. 
            onSubmit={(expense) => {   
                //props.dispatch(addExpense(expense));    // this line was used BEFORE setting up test case
                props.onSubmit(expense);                  // this line was used AFTER  setting up test case in conjuction with 
                                                          // mapDispatchToProps
                //  push allows you to programmatically change pages. Argument should be a path or route
                props.history.push("/");      // Redirect using props.history from AppRouter available props
                                              // Note, this uses browser routing. It does not through a page refresh      
                                                   
            }}                        
        />
    </div>
);

// The goal of mapDispatchToProps is to return an object that Maps your dispatch call to a prop vs calling dispatch inline.
// This allows you to abstract away dispatch from the compoment itself.
// Instead of having following inline code:    props.dispatch(addExpense(expense))    
// in the onSubmit prop used to render ExpenseForm above, we create the mapDispatchToProps function and 
// use following:                              props.onSubmit(expense)
// See the BEFORE and AFTER changes noted in the props of the  <ExpenseForm /> compoment above.
const mapDispatchToProps = (dispatch) => ({
        onSubmit: (expense)  => dispatch(addExpense(expense))
});


export default connect(undefined, mapDispatchToProps)(AddExpensePage);