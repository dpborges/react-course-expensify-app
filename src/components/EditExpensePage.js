import React from 'react';
import { connect } from 'react-redux';
import ExpenseForm from './ExpenseForm';
import { editExpense, startRemoveExpense  } from '../actions/expenses';


// See AddExpensePageOrig.js first then the AddExpensePage.js compoment.
// What was done there is same process we took with EditExpensePage

// This component gets called by AppRouter with route /edit/:id   passing in id
// as prop called props.match.params.id.
export class EditExpensePage extends React.Component  { 

    onSubmit =  (expense) => {
        this.props.editExpense(this.props.expense.id, expense);
        this.props.history.push("/");
    };

    onRemove = () => {
        this.props.startRemoveExpense({ id: this.props.expense.id });
        this.props.history.push("/");
    };

    
    // console.log('The ID: ' , props.match.params.id);
    render() {
        // console.log('EXPENSE PAGE PROPS: ' , this.props);
        return (
            <div>
                <ExpenseForm
                    expense={this.props.expense}
                    onSubmit={this.onSubmit}
                />
                <button 
                    onClick={this.onRemove}>
                    Remove
                </button>
            </div>
        );
    }
};

//Uses implicit return
const mapDispatchToProps = (dispatch, ownProps) => ({
    editExpense: (id, expense)  => dispatch(editExpense(id, expense)),
    startRemoveExpense: ( data ) => dispatch(startRemoveExpense(data))
});


//Uses explicit return
// Find the single expense we want to edit and pass via props to component
const mapStateToProps = (state, props) => {
    return {
        expense: state.expenses.find((expense) => {
            return expense.id === props.match.params.id;
        })
    }
}

export default connect(mapStateToProps, mapDispatchToProps)(EditExpensePage);