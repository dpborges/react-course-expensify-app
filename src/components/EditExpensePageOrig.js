import React from 'react';
import { connect } from 'react-redux';
import ExpenseForm from './ExpenseForm';
import { editExpense } from '../actions/expenses';
import { removeExpense } from '../actions/expenses';

// See AddExpensePageOrig.js first then the AddExpensePage.js compoment.
// What was done there is same process we took with EditExpensePage

// This component gets called by AppRouter for route /edit/:id   passing in id
// as prop called props.match.params.id.

const EditExpensePage = (props) => { 
    console.log('EXPENSE PAGE PROPS: ' , props);
    console.log('The ID: ' , props.match.params.id);
    return (
        <div>
            <ExpenseForm
                expense={props.expense}
                onSubmit={(expense) => {
                    props.dispatch(editExpense(props.expense.id, expense));
                    props.history.push("/");
                }}
            />
            
            <button 
                onClick={() => {
                    props.dispatch(removeExpense({ id: props.expense.id }));
                    props.history.push("/");
            }}>
                Remove Expense
            </button>
        </div>
    );
};

// Find the single expense we want to edit and pass via props to component
const mapStateToProps = (state, props) => {
    return {
        expense: state.expenses.find((expense) => {
            return expense.id === props.match.params.id;
        })
    }
}

export default connect(mapStateToProps)(EditExpensePage);