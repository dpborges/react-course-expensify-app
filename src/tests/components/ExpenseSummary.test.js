import React from 'react';
import { shallow } from 'enzyme';
import expenses from '../fixtures/expenses';
import { ExpensesSummary } from '../../components/ExpensesSummary';


// Note when testing readonly functional components, you can pass in props from your test case
test('should correctly render ExpenseSummary with one expense', () => {
    const wrapper = shallow(<ExpensesSummary expenseCount={1} expensesTotal={235} />);
    expect(wrapper).toMatchSnapshot();
});



test('should correctly render ExpenseSummary with multiple expenses', () => {
    const wrapper = shallow(<ExpensesSummary expenseCount={3} expensesTotal={3510050}  />);  
    expect(wrapper).toMatchSnapshot();
});