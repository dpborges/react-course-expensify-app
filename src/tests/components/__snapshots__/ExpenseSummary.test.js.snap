// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should correctly render ExpenseSummary with multiple expenses 1`] = `
<div>
  <p>
    Viewing 
    3
     
    expenses
     totalling 
    $35,100.50
  </p>
</div>
`;

exports[`should correctly render ExpenseSummary with one expense 1`] = `
<div>
  <p>
    Viewing 
    1
     
    expense
     totalling 
    $2.35
  </p>
</div>
`;
